unit/testing-utils/require_type.cpp

```diff
@@ -78,3 +93,109 @@ code_typet::parametert require_type::require_parameter(
+      REQUIRE(param.type_variable().get_identifier()==expected.description);
+      return true;
+    case require_type::type_parameter_kindt::Inst:
+      REQUIRE(is_java_generic_inst_parameter((param)));
+      REQUIRE(param.subtype()==symbol_typet(expected.description));
+      return true;
+  }
+  // Should be unreachable...
+  REQUIRE(false);
+  return false;
+}
+
+
+/// Verify a given type is a java_generic_type, optionally checking
+/// that it's associated type variables match a given set of identifiers.
+/// Expected usage is something like this:
+///
+/// require_java_generic_type(type,
+///                          {{Inst,"java::java.lang.Integer"},{Var,"T"}})
+///
+/// \param type The type to check
+/// \param type_expectations An optional set of type variable kinds
+/// and identifiers which should be expected as the type parameters of the
+/// given generic type.
+/// \return The given type, cast to a java_generic_typet
+java_generic_typet require_type::require_java_generic_type(
+  const typet &type,
+  const optionalt<
+    std::initializer_list<require_type::expected_type_parametert>
+  > &type_expectations)
+{
+  REQUIRE(is_java_generic_type(type));
+  const java_generic_typet &generic_type=to_java_generic_type(type);
+  if(type_expectations)
+  {
+    const java_generic_typet::generic_type_variablest &generic_type_vars=
+      generic_type.generic_type_variables();
+    REQUIRE(generic_type_vars.size()==type_expectations->size());
+    REQUIRE(
+      std::equal(
+        generic_type_vars.begin(),
+        generic_type_vars.end(),
+        type_expectations->begin(),
+        require_java_generic_parametert_expectation));
+  }
+
+  return to_java_generic_type(type);
+}
+
+/// Verify a given type is a java_generic_parameter, optionally checking
+/// that it's associated type variables match a given set of expectations.
+/// Expected usage is something like this:
+///
+/// require_java_generic_parameter(parameter, {Inst,"java::java.lang.Integer"})
+///
+/// or
+///
+/// require_java_generic_parameter(parameter, {Var,"T"})
+///
+/// \param type The type to check
+/// \param type_expectation An optional description of the identifiers/kinds
+/// which / should be expected as the type parameter of the generic parameter.
+/// \return The given type, cast to a java_generic_parametert
+java_generic_parametert require_type::require_java_generic_parameter(
+  const typet &type,
+  const optionalt<require_type::expected_type_parametert> &type_expectation)
+{
+  REQUIRE(is_java_generic_parameter(type));
+  const java_generic_parametert &generic_param=to_java_generic_parameter(type);
+  if(type_expectation)
+  {
+    REQUIRE(
+      require_java_generic_parametert_expectation(
+        generic_param,
+        type_expectation.value()));
+  }
+
+  return generic_param;
+}
+
+const typet &require_type::require_java_non_generic_type(
+  const typet &type,
+  const optionalt<irep_idt> &expect_type)
```

Comment:
This parameter seems a bit strange - I'd prefer if it took a `symbol_typet` or maybe just leave checking the actual type to a separate helper function.

Reply:
Yeah, that is I think a hangover from an earlier usage of this helper. Fixed now, including adding the missing doxygen.